sec1_intro_and_setup/test_app/config/routes.rb

```diff
@@ -0,0 +1,3 @@
+Rails.application.routes.draw do
```

Comment:
Whats this file for?

Reply:
This file is used to set up routes as you build your app.

Reply:
What are `routes`?

Reply:
Routes are connectors around the Controllers.  Routes take incoming requests to the server and direct them to the correct Controller. Then, the Controller digest the requests and produce appropriate output to the Model and the View.

Reply:
A user enters an URL to go to the homepage of Facebook through the browser. Then the router decides the request be handled by a controller and the action is to go to the homepage. The controller then renders back HTML to the user as a response.

Reply:
You are skimming over the details.
Go into the details.

User input URL. Url -> magically goes into controller? which controller? what is action? which action will be called? how does the controller "render" html? 

I still don't understand how this works based on your explanation

Reply:
User request URL `pages/home`. In `routes.rb`, 
```
Rails.application.routes.draw do
  get "/pages/home" => "pages#show"
end
```
The router then decides the request be handled by `pages_controller` and the controller action as `show`. You can find the `pages_controller` in `app/controllers/pages_controller.rb` (if you created the `pages_controller`)
```
class PagesController < ApplicationController
  def show
    render template: "pages/#home"
  end
end
```
Note that we tell the `show` action to render a page template. You can find the page template at `app/views/pages`, where the controller renders back html (`app/views/pages/home.html.erb`) as a response to user.

Reply:
1. User request URL ... <= Why when user request, it is a `GET` ?
2. How does the router decide the request be handled by which controller and which action?

Reply:
1. When user request URL, we do not intend to let the user to change the state of server, so we use `GET` which is used to read/retrieve a representation of a resource. While POST, PUT, DELETE, PATCH can do request but they can modify the state of server, for example:
- `POST` is used to create sub-resources of its parent (Eg. `http://.../comments/1` where `comments` is parent and `1` is sub-resources.
- `PUT` is used to replace resources if request body exist, else if the request body is not found it will create a new resources. To update, user needs to specify the resource ID(eg. `PUT http://.../comments/1`), otherwise `PUT http://.../comments`. 
(NOTE: POST is used to create resources without the client knowing the URL while PUT is used to create resources at a URL known by client)
- `DELETE` is used to delete a resource specified by the URL.
- `PATCH` is used to update resources. It is used to modify the current resources to produce a new version.

2. Yes, you can have `xxx_controller` with action `yyy` be triggered. You can create a module for the action `yyy` in `xxx_controller` and include the module in `pages_controller`
```
module xxx_module
   def yyy
       ...
   end

class PagesController < ApplicationController
   include xxx_module
```